File: 0380/solution.py
from random import choice


class RandomizedSet:

    def __init__(self):
        self.items = []
        self.locations = {}

    def insert(self, val: int) -> bool:
        if val in self.locations:
            return False
        self.items.append(val)
        self.locations[val] = len(self.items) - 1
        return True

    def remove(self, val: int) -> bool:
        if val not in self.locations:
            return False
        to_remove = self.locations[val]
        last_item = self.items[-1]
        self.locations[last_item] = to_remove
        self.items[to_remove], self.items[-1] = self.items[-1], self.items[to_remove]
        self.items.pop()
        self.locations.pop(val)
        return True

    def getRandom(self) -> int:
        return choice(self.items)

File: 0380/solution.ts
class RandomizedSet {
  locs: Map<number, number>;
  nums: Array<number>;
  constructor() {
    this.locs = new Map();
    this.nums = [];
  }

  insert(val: number): boolean {
    if (this.locs.has(val)) {
      return false;
    }
    this.locs.set(val, this.nums.length);
    this.nums.push(val);
    return true;
  }

  remove(val: number): boolean {
    if (!this.locs.has(val)) {
      return false;
    }
    const lastNum = this.nums[this.nums.length - 1];
    const removeLoc = this.locs.get(val);
    this.locs.set(lastNum, removeLoc!);
    this.locs.delete(val);
    this.nums[removeLoc!] = lastNum;
    this.nums.pop();
    return true;
  }

  getRandom(): number {
    const randomIndex = Math.floor(Math.random() * this.nums.length);
    return this.nums[randomIndex];
  }
}
